File: src/main/scala/com/vc/hard/L305.scala
package com.vc.hard

object L305 {
  def numIslands2(m: Int, n: Int, pos: Array[Array[Int]]): List[Int] = {
    val pn = pos.length
    val res = new Array[Int](pn)
    val rootArr = new Array[Int](m * n)
    rootArr.indices.foreach(i => {
      rootArr(i) = -1
    })
    def findRoot(root: Int): Int = {
      if(root != rootArr(root)) {
        return findRoot(rootArr(root))
      }
      else return root
    }
    var count = 0
    val dirs = Array((0, 1), (1, 0), (0, -1), (-1, 0))
    pos.indices.foreach(i => {
      val px = pos(i)(0)
      val py = pos(i)(1)

      val land = px * n + py
      if(rootArr(land) == -1) {
        rootArr(land) = land
        var root = land
        count += 1

        dirs.foreach(dir => {
          val xNew = px + dir._1
          val yNew = py + dir._2
          val landNb = xNew * n + yNew
          if(xNew >= 0 && xNew < m && yNew >= 0 && yNew < n && rootArr(landNb) != -1) {
            val rootNb = findRoot(landNb)
            if(rootNb != root) {
              rootArr(rootNb) = root
              count -= 1
            }
          }
        })
      }
      res(i) = count
    })
    res.toList
  }
}

File: src/main/java/com/vc/hard/L305.java
package com.vc.hard;

import java.util.*;

class L305 {
    public List<Integer> numIslands2(int m, int n, int[][] positions) {
        List<Integer> res = new ArrayList<>();
        int[] rootArr = new int[m * n];
        Arrays.fill(rootArr, -1);
        int[][] dirs = {{0,1},{0,-1},{1,0},{-1,0}};
        int count = 0;
        for(int i = 0; i < positions.length; i++) {
            int px = positions[i][0];
            int py = positions[i][1];
            int root = px * n + py;
            if(rootArr[root] == -1) {
                count++;
                rootArr[root] = root;
                for(int[] dir: dirs) {
                    int xNew = px + dir[0];
                    int yNew = py + dir[1];
                    int rootNb = xNew * n + yNew;
                    if(xNew >= 0 && xNew < m && yNew >= 0 && yNew < n && rootArr[rootNb] != -1){
                        rootNb = findRoot(rootNb, rootArr);
                        if(rootNb != root) {
                            count--;
                            rootArr[rootNb] = root;
                        }
                    }
                }
            }
            res.add(count);
        }
        return res;
    }

    private int findRoot(int root, int[] rootArr) {
        if(rootArr[root] == root) return root;
        else return findRoot(rootArr[root], rootArr);
    }
}